editor/components/colors/with-colors.js

```diff
@@ -1,91 +1,129 @@
-					* a new function reference is returned (even if arguments have not changed).
-					* This would make our memoized chain useless.
-					*/
-					this.memoizedGetColor = memoize( this.memoizedGetColor, { maxSize: 1 } );
-					this.memoizedSetColor = memoize( this.memoizedSetColor, { maxSize: 1 } );
-				}
+			( WrappedComponent ) => {
+				return class extends Component {
+					constructor( props ) {
+						super( props );
+
+						this.setters = this.createSetters();
 
-				memoizedGetColor( colors ) {
-					return memoize(
-						( colorName, customColorValue, colorContext ) => {
-							return {
-								name: colorName,
-								class: getColorClass( colorContext, colorName ),
-								value: getColorValue( colors, colorName, customColorValue ),
-							};
-						}
-					);
-				}
+						this.state = {};
+					}
 
-				memoizedSetColor( setAttributes, colors ) {
-					return memoize(
-						( colorNameAttribute, customColorAttribute ) => {
-							return setColorValue( colors, colorNameAttribute, customColorAttribute, setAttributes );
-						}
-					);
-				}
+					createSetters() {
+						return reduce( colorMap, ( settersAcc, colorContext, colorAttributeName ) => {
+							const ufColorAttrName = upperFirst( colorAttributeName );
+							const customColorAttributeName = `custom${ ufColorAttrName }`;
+							settersAcc[ `set${ ufColorAttrName }` ] = this.createSetColor( colorAttributeName, customColorAttributeName );
+							return settersAcc;
+						}, {} );
+					}
 
-				render() {
-					return (
-						<WrappedComponent
-							{ ...{
-								...this.props,
-								colors: undefined,
-								...mapGetSetColorToProps(
-									this.memoizedGetColor( this.props.colors ),
-									this.memoizedSetColor( this.props.setAttributes, this.props.colors ),
-									this.props
-								),
-							} }
-						/>
-					);
-				}
-			};
-		},
-	] ),
-	'withColors'
-);
+					createSetColor( colorAttributeName, customColorAttributeName ) {
+						return 	( ( colorValue ) => {
+							const colorObj = find( this.props.colors, { color: colorValue } );
+							this.props.setAttributes( {
+								[ colorAttributeName ]: colorObj && colorObj.name ? colorObj.name : undefined,
+								[ customColorAttributeName ]: colorObj && colorObj.name ? undefined : colorValue,
+							} );
+						} ).bind( this );
+					}
+
+					static getDerivedStateFromProps( { attributes, colors }, prevState ) {
+						return reduce( colorMap, ( newState, colorContext, colorAttributeName ) => {
+							const colorName = attributes[ colorAttributeName ];
+							const colorValue = getColorValue(
+								colors,
+								colorName,
+								attributes[ `custom${ upperFirst( colorAttributeName ) }` ]
+							);
+							const prevColorObject = prevState[ colorAttributeName ];
+							const prevColorValue = get( prevColorObject, [ 'value' ] );
+							if ( prevColorValue === colorValue && prevColorValue ) {
```

Comment:
For what case do we need `&& prevColorValue` ? Can we have a unit test?

Reply:
It is a trivial case and the condition is executed all the time, without `&& prevColorValue` during the first execution, both prevColorValue and colorValue are undefined. So we would always execute `newState[ colorAttributeName ] = prevColorObject;` and never compute the object. I will add this as a comment to make clear why `&& prevColorValue`  is needed.